flink-tests/src/test/java/org/apache/flink/test/checkpointing/utils/FailingSource.java

```diff
@@ -0,0 +1,155 @@
+	implements ListCheckpointed<Integer>, CheckpointListener {
+
+	/**
+	 * Function to generate and emit the test events (and watermarks if required).
+	 */
+	@FunctionalInterface
+	public interface EventEmittingGenerator extends Serializable {
+		void emitEvent(SourceContext<Tuple2<Long, IntType>> ctx, int eventSequenceNo);
+	}
+
+	private static final long INITIAL = Long.MIN_VALUE;
+	private static final long STATEFUL_CHECKPOINT_COMPLETED = Long.MIN_VALUE;
+
+	@Nonnull
+	private final EventEmittingGenerator eventEmittingGenerator;
+	private final int expectedEmitCalls;
+	private final int failureAfterNumElements;
+	private final boolean usingProcessingTime;
+	private final AtomicLong checkpointStatus;
+
+	private int emitCallCount;
+	private volatile boolean running;
+
+	public FailingSource(
+		@Nonnull EventEmittingGenerator eventEmittingGenerator,
+		@Nonnegative int numberOfGeneratorInvocations) {
+		this(eventEmittingGenerator, numberOfGeneratorInvocations, TimeCharacteristic.EventTime);
+	}
+
+	public FailingSource(
+		@Nonnull EventEmittingGenerator eventEmittingGenerator,
+		@Nonnegative int numberOfGeneratorInvocations,
+		@Nonnull TimeCharacteristic timeCharacteristic) {
+		this.eventEmittingGenerator = eventEmittingGenerator;
+		this.running = true;
+		this.emitCallCount = 0;
+		this.expectedEmitCalls = numberOfGeneratorInvocations;
+		this.failureAfterNumElements = numberOfGeneratorInvocations / 2;
+		this.checkpointStatus = new AtomicLong(INITIAL);
+		this.usingProcessingTime = timeCharacteristic == TimeCharacteristic.ProcessingTime;
+	}
+
+	@Override
+	public void open(Configuration parameters) {
+		// non-parallel source
+		assertEquals(1, getRuntimeContext().getNumberOfParallelSubtasks());
+	}
+
+	@Override
+	public void run(SourceContext<Tuple2<Long, IntType>> ctx) throws Exception {
+
+		final RuntimeContext runtimeContext = getRuntimeContext();
+		// detect if this task is "the chosen one" and should fail (via subtaskidx), if it did not fail before (via attempt)
+		final boolean failThisTask =
+			runtimeContext.getAttemptNumber() == 0 && runtimeContext.getIndexOfThisSubtask() == 0;
+
+		// we loop longer than we have elements, to permit delayed checkpoints
+		// to still cause a failure
+		while (running) {
+
+			// the function failed before, or we are in the elements before the failure
+			synchronized (ctx.getCheckpointLock()) {
+				eventEmittingGenerator.emitEvent(ctx, emitCallCount++);
+				running &= (emitCallCount < expectedEmitCalls);
+			}
+
+			if (emitCallCount < failureAfterNumElements) {
+				Thread.sleep(1);
+			} else if (failThisTask && emitCallCount == failureAfterNumElements) {
+				// wait for a pending checkpoint that fulfills our requirements if needed
+				while (checkpointStatus.get() != STATEFUL_CHECKPOINT_COMPLETED) {
+					Thread.sleep(1);
+				}
+				throw new Exception("Artificial Failure");
+			}
+		}
+
+		if (usingProcessingTime) {
```

Comment:
why are we only looping at the end if using processing time? This was looping indefinitely before to allow checkpoints to still go through.

Reply:
Because in event time this is deterministic because everything gets flushed with a max-watermark and we have a graceful shutdown.